File: mobile/android/TARSCompanion/app/src/main/java/world/meeet/tars/net/PairingClient.kt
/*
 * PairingClient — OkHttp driver for the L5 pairing handshake.
 *
 * Mirrors the iOS PairingClient surface 1:1 so the contract tests
 * can compare both sides against a single source of truth.
 */

package world.meeet.tars.net

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import world.meeet.tars.TARSCompanion
import java.util.concurrent.TimeUnit

data class PairingHostBegin(
    val pairID: String,
    val acceptToken: String,
    val hostID: String,
    val hostFingerprint: String,
    val hostPublicKey: String,
    val expiresAt: Double,
    val traceID: String?,
)

enum class PairingState(val raw: String) {
    PENDING("pending"),
    LINKED("linked"),
    EXPIRED("expired"),
    REJECTED("rejected"),
    UNKNOWN("unknown");

    companion object {
        fun fromRaw(raw: String?): PairingState =
            values().firstOrNull { it.raw == raw } ?: UNKNOWN
    }
}

data class PairingStatus(
    val pairID: String,
    val state: PairingState,
    val deviceID: String?,
    val hostFingerprint: String?,
)

sealed class PairingClientError(message: String) : Exception(message) {
    object InvalidURL : PairingClientError("invalid url")
    class Http(val code: Int, val body: String) : PairingClientError("HTTP $code")
    class Malformed(val reason: String) : PairingClientError("malformed: $reason")
    class Network(val reason: String) : PairingClientError("network: $reason")
}

class PairingClient(
    private val baseURL: String,
    private val client: OkHttpClient = defaultClient(),
) {
    fun begin(clientEphemeralKeyBase64: String): PairingHostBegin {
        val body = JSONObject()
            .put("client_epk", clientEphemeralKeyBase64)
            .put("kind", TARSCompanion.DEVICE_KIND)
            .toString()
        val request = Request.Builder()
            .url(resolve("/api/pairing/begin"))
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()
        val data = perform(request)
        return decodeBegin(data)
    }

    fun pollStatus(pairID: String): PairingStatus {
        val url = resolve("/api/pairing/status").toHttpUrlOrNull()
            ?: throw PairingClientError.InvalidURL
        val request = Request.Builder()
            .url(url.newBuilder().addQueryParameter("pair_id", pairID).build())
            .get()
            .build()
        val data = perform(request)
        return decodeStatus(data)
    }

    private fun perform(request: Request): String {
        try {
            client.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw PairingClientError.Http(response.code, body)
                }
                return body
            }
        } catch (t: PairingClientError) {
            throw t
        } catch (t: Throwable) {
            throw PairingClientError.Network(t.message ?: t.toString())
        }
    }

    private fun resolve(path: String): String {
        val base = baseURL.trimEnd('/')
        return base + path
    }

    companion object {
        fun defaultClient(): OkHttpClient =
            OkHttpClient.Builder()
                .connectTimeout(8, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .build()

        fun decodeBegin(raw: String): PairingHostBegin {
            val json = parseObject(raw)
            return PairingHostBegin(
                pairID = json.requireString("pair_id"),
                acceptToken = json.requireString("accept_token"),
                hostID = json.requireString("host_id"),
                hostFingerprint = json.requireString("host_fingerprint"),
                hostPublicKey = json.requireString("host_public_key"),
                expiresAt = json.optDouble("expires_at"),
                traceID = json.optStringOrNull("trace_id"),
            )
        }

        fun decodeStatus(raw: String): PairingStatus {
            val json = parseObject(raw)
            return PairingStatus(
                pairID = json.requireString("pair_id"),
                state = PairingState.fromRaw(json.optStringOrNull("state")),
                deviceID = json.optStringOrNull("device_id"),
                hostFingerprint = json.optStringOrNull("host_fingerprint"),
            )
        }

        private fun parseObject(raw: String): JSONObject = try {
            JSONObject(raw)
        } catch (t: Throwable) {
            throw PairingClientError.Malformed("not a JSON object: ${t.message}")
        }

        private fun JSONObject.requireString(key: String): String {
            if (!has(key)) throw PairingClientError.Malformed("missing $key")
            return optString(key, "").ifEmpty {
                throw PairingClientError.Malformed("empty $key")
            }
        }

        private fun JSONObject.optStringOrNull(key: String): String? {
            if (!has(key) || isNull(key)) return null
            val v = optString(key, "")
            return if (v.isEmpty()) null else v
        }
    }
}

private fun String.toHttpUrlOrNull(): okhttp3.HttpUrl? = okhttp3.HttpUrl.parse(this)

File: mobile/ios/TARSCompanion/Sources/TARSCompanion/PairingClient.swift
// PairingClient — URLSession driver for the L5 pairing handshake.
//
// Implements exactly three calls used by the L1 slice:
//
//   POST /api/pairing/begin     — submit our `client_epk`.
//   GET  /api/pairing/status    — poll for {pending → linked|expired|rejected}.
//   GET  /api/pairing/identity  — sanity-check the host's fingerprint.
//
// No `Codable` magic: each response is decoded by hand so a stray
// schema drift fails loud instead of silently dropping a field.

import Foundation

public struct PairingHostBegin: Equatable, Sendable {
    public let pairID: String
    public let acceptToken: String
    public let hostID: String
    public let hostFingerprint: String
    public let hostPublicKey: String
    public let expiresAt: Double
    public let traceID: String?
}

public enum PairingState: String, Sendable {
    case pending
    case linked
    case expired
    case rejected
    case unknown
}

public struct PairingStatus: Equatable, Sendable {
    public let pairID: String
    public let state: PairingState
    public let deviceID: String?
    public let hostFingerprint: String?
}

public enum PairingClientError: Error, Equatable {
    case invalidURL
    case http(Int, String)
    case malformed(String)
    case network(String)
}

public actor PairingClient {
    public let baseURL: URL
    private let session: URLSession

    public init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    // MARK: – /api/pairing/begin

    public func begin(clientEphemeralKey base64PublicKey: String) async throws -> PairingHostBegin {
        let body: [String: Any] = [
            "client_epk": base64PublicKey,
            "kind": TARSCompanion.deviceKind,
        ]
        let data = try await postJSON(path: "/api/pairing/begin", body: body)
        return try Self.decodeBegin(data)
    }

    // MARK: – /api/pairing/status

    public func pollStatus(pairID: String) async throws -> PairingStatus {
        var url = try resolve("/api/pairing/status")
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false) ?? URLComponents()
        components.queryItems = [URLQueryItem(name: "pair_id", value: pairID)]
        guard let composed = components.url else {
            throw PairingClientError.invalidURL
        }
        url = composed
        let data = try await getJSON(url: url)
        return try Self.decodeStatus(data)
    }

    // MARK: – decoders

    static func decodeBegin(_ data: Data) throws -> PairingHostBegin {
        let json = try parseObject(data)
        guard let pairID = json["pair_id"] as? String,
              let acceptToken = json["accept_token"] as? String,
              let hostID = json["host_id"] as? String,
              let hostFingerprint = json["host_fingerprint"] as? String,
              let hostPublicKey = json["host_public_key"] as? String,
              let expiresAt = json["expires_at"] as? Double else {
            throw PairingClientError.malformed("missing field in /pairing/begin response")
        }
        return PairingHostBegin(
            pairID: pairID,
            acceptToken: acceptToken,
            hostID: hostID,
            hostFingerprint: hostFingerprint,
            hostPublicKey: hostPublicKey,
            expiresAt: expiresAt,
            traceID: json["trace_id"] as? String
        )
    }

    static func decodeStatus(_ data: Data) throws -> PairingStatus {
        let json = try parseObject(data)
        guard let pairID = json["pair_id"] as? String,
              let stateRaw = json["state"] as? String else {
            throw PairingClientError.malformed("missing field in /pairing/status response")
        }
        return PairingStatus(
            pairID: pairID,
            state: PairingState(rawValue: stateRaw) ?? .unknown,
            deviceID: json["device_id"] as? String,
            hostFingerprint: json["host_fingerprint"] as? String
        )
    }

    // MARK: – HTTP plumbing

    private func postJSON(path: String, body: [String: Any]) async throws -> Data {
        var request = URLRequest(url: try resolve(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await perform(request: request)
    }

    private func getJSON(url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        return try await perform(request: request)
    }

    private func perform(request: URLRequest) async throws -> Data {
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw PairingClientError.malformed("non-HTTP response")
            }
            guard (200..<300).contains(http.statusCode) else {
                let body = String(data: data, encoding: .utf8) ?? ""
                throw PairingClientError.http(http.statusCode, body)
            }
            return data
        } catch let err as PairingClientError {
            throw err
        } catch {
            throw PairingClientError.network(error.localizedDescription)
        }
    }

    private func resolve(_ path: String) throws -> URL {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw PairingClientError.invalidURL
        }
        return url
    }

    private static func parseObject(_ data: Data) throws -> [String: Any] {
        do {
            guard let any = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw PairingClientError.malformed("response is not a JSON object")
            }
            return any
        } catch let err as PairingClientError {
            throw err
        } catch {
            throw PairingClientError.malformed("invalid JSON: \(error.localizedDescription)")
        }
    }
}
